tools/indexer/cmd/start.go

```diff
@@ -0,0 +1,336 @@
+	launchAsyncValidatorTasks(ctx, c, db, httpc, ei, beaconBase, logger)
+
+	saveBlockProgress(db, nextBN, logger)
+	return nextBN
+}
+
+func processBidsForBlock(ctx context.Context, db *database.DB, httpc *retryablehttp.Client, relays []relay.Row, ei *beacon.ExecInfo, logger *slog.Logger) {
+
+	// Fetch and store bid data from all relays
+	totalBids := 0
+	successfulRelays := 0
+	mainContextCanceled := false
+	const batchSize = 500
+	for _, rr := range relays {
+		if ctx.Err() != nil {
+			logger.Warn("main context canceled, stopping relay processing")
+			break
+		}
+
+		bids, err := relay.FetchBuilderBlocksReceived(ctx, httpc, rr.URL, ei.Slot)
+		if err != nil {
+			logger.Error("[RELAY] failed to fetch bids", "relay_id", rr.ID, "url", rr.URL, "error", err)
+			continue
+		}
+
+		relayBids := 0
+		batch := make([]database.BidRow, 0, batchSize)
+
+		for _, bid := range bids {
+			// Check if main context is still valid
+			if ctx.Err() != nil {
+				logger.Warn("[BIDS] main context canceled, stopping bid insertion")
+				mainContextCanceled = true
+				break
+			}
+
+			if row, ok := relay.BuildBidInsert(ei.Slot, rr.ID, bid); ok {
+				batch = append(batch, row)
+
+				if len(batch) >= batchSize {
+					insCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+					if err := db.InsertBidsBatch(insCtx, batch); err != nil {
+
+						logger.Error("[DB]batch insert failed", "slot", ei.Slot, "relay_id", rr.ID, "count", len(batch), "error", err)
+					} else {
+						relayBids += len(batch)
+					}
+					cancel()
+					batch = batch[:0]
+				}
+			}
+		}
+
+		// final flush
+		if len(batch) > 0 {
+			flushCtx, flushCancel := context.WithTimeout(context.Background(), 5*time.Second)
+			if err := db.InsertBidsBatch(flushCtx, batch); err != nil {
+				logger.Error("[DB] batch insert failed", "slot", ei.Slot, "relay_id", rr.ID, "count", len(batch), "error", err)
+			} else {
+				relayBids += len(batch)
+			}
+			flushCancel()
+		}
+
+		if mainContextCanceled {
+			break
+		}
+
+		if relayBids > 0 {
+			logger.Info("[BIDS] bids collected", "relay_id", rr.ID, "count", relayBids)
+			totalBids += relayBids
+			successfulRelays++
+		}
+	}
+	logger.Info("[BIDS] summary", "block", ei.BlockNumber, "total_bids", totalBids, "successful_relays", successfulRelays)
+
+}
+
+func saveBlockProgress(db *database.DB, blockNum int64, logger *slog.Logger) {
+	gctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	if err := db.SaveLastBlockNumber(gctx, blockNum); err != nil {
+		logger.Error("[PROGRESS] failed to save block number", "block", blockNum, "error", err)
+	} else {
+		logger.Info("[PROGRESS] advanced to block", "block", blockNum)
+	}
+
+}
+
+func launchAsyncValidatorTasks(ctx context.Context, c *cli.Context, db *database.DB, httpc *retryablehttp.Client, ei *beacon.ExecInfo, beaconBase string, logger *slog.Logger) { // Async validator pubkey fetch
+	if ei.ProposerIdx != nil {
+		go func(slot int64, proposerIdx int64) {
```

Comment:
So the first go routine is supposed to populate the Validator public key which the second routine is using? What is the point of doing it in separate go routines then? The second one will fail till first one completes. Not sure if I understand this logic.